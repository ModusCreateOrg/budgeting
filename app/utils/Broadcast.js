// @flow

type Subscription = (currentState: mixed) => void;

/**
 * Broadcast
 *
 * Event emitter which provides a generic way to pass values to children
 * using React's context.
 *
 * We have to use this because we can't send data to children directly in the context.
 * If we do so, and `shouldComponentUpdate` returns false somewhere in the child tree,
 * then it's children won't get the updated context when it changes.
 *
 * To be able to send data, context should not change. Instead, context is used
 * to notify components about state changes.
 */
class Broadcast {
  currentState: mixed;
  subscriptions: Subscription[];

  constructor(initialState: mixed): void {
    this.currentState = initialState;
    this.subscriptions = [];
  }

  getState = (): mixed => this.currentState;

  setState = (newState: mixed): void => {
    this.currentState = newState;

    this.subscriptions.forEach(subscription => subscription(this.currentState));
  };

  subscribe = (subscription: Subscription): (() => void) => {
    this.subscriptions.push(subscription);

    // return `unsubscribe` function
    return () => {
      this.subscriptions = this.subscriptions.filter(item => item !== subscription);
    };
  };
}

export default Broadcast;
