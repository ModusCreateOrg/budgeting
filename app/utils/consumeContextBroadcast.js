// @flow
import * as React from 'react';
import PropTypes from 'prop-types';
import getDisplayName from 'utils/getDisplayName';
import type { HigherOrderComponent } from 'types';
import Broadcast from './Broadcast';

type ProvidedProps = {
  [broadcastName: string]: mixed,
};

type ConsumeContextBroadcastState = {
  broadcastState: mixed,
};

/**
 * Returns a HOC you can use to enable a component to receive data
 * from a form data broadcaster in context.
 */
const consumeContextBroadcast = (
  broadcastName: string
): HigherOrderComponent<{}, ProvidedProps> => WrappedComponent => {
  class ConsumeContextBroadcast extends React.Component<{}, ConsumeContextBroadcastState> {
    static contextTypes = {
      [broadcastName]: PropTypes.object,
    };

    state = {
      broadcastState: null,
    };

    componentWillMount() {
      const broadcast = this.context[broadcastName];

      if (!(broadcast instanceof Broadcast)) {
        throw new Error('consumeContextBroadcast must be rendered in the context of a broadcast with the right name');
      }

      // set initial state from the broadcast
      const initialState = broadcast.getState();
      this.updateState(initialState);
    }

    componentDidMount() {
      const broadcast = this.context[broadcastName];

      // subscribe to the form data broadcast
      this.unsubscribe = broadcast.subscribe(this.updateState);
    }

    componentWillUnmount() {
      this.unsubscribe();
    }

    unsubscribe: () => void;

    updateState = broadcastState => {
      this.setState({ broadcastState: broadcastState });
    };

    render() {
      const { broadcastState } = this.state;
      const props = {
        ...this.props,
        [broadcastName]: broadcastState,
      };

      return <WrappedComponent {...props} />;
    }
  }

  ConsumeContextBroadcast.displayName = `ConsumeContextBroadcast(${getDisplayName(WrappedComponent)})`;
  return ConsumeContextBroadcast;
};

export default consumeContextBroadcast;
