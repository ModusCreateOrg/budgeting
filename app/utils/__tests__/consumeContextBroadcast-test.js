import * as React from 'react';
import PropTypes from 'prop-types';
import renderer from 'react-test-renderer';
import Broadcast from '../Broadcast';
import consumeContextBroadcast from '../consumeContextBroadcast';

beforeAll(() => {
  // A console.error is expected here, saying that ErrorBoundary will handle an error.
  // We mock this to see a clean output in the test runs.
  console.error = jest.fn();
});

describe('consumeContextBroadcast', () => {
  it('should return a HOC that raises an error if no broadcast exists in context', () => {
    const hoc = consumeContextBroadcast('testBroadcast');
    const WrappedComponent = () => <div>test</div>;
    const GeneratedComponent = hoc(WrappedComponent);

    class ErrorBoundary extends React.Component {
      // eslint-disable-next-line class-methods-use-this
      componentDidCatch(error) {
        expect(error.message).toEqual(
          'consumeContextBroadcast must be rendered in the context of a broadcast with the right name'
        );
      }

      render() {
        return this.props.children;
      }
    }

    renderer.create(
      <ErrorBoundary>
        <GeneratedComponent />
      </ErrorBoundary>
    );
  });

  it('should return a HOC that renders correctly if the right broadcast exists in context', () => {
    const hoc = consumeContextBroadcast('testBroadcast');
    const WrappedComponent = () => <div>test</div>;
    const GeneratedComponent = hoc(WrappedComponent);

    // eslint-disable-next-line react/no-multi-comp
    class ContextProvider extends React.Component {
      static childContextTypes = {
        testBroadcast: PropTypes.object,
      };

      getChildContext() {
        return { testBroadcast: new Broadcast('initialState') };
      }

      render() {
        return this.props.children;
      }
    }

    const tree = renderer
      .create(
        <ContextProvider>
          <GeneratedComponent />
        </ContextProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it(`should return a HOC that sets a Broadcast state in the context`, () => {
    const hoc = consumeContextBroadcast('testBroadcast');
    const broadcast = new Broadcast('initialState');

    // eslint-disable-next-line react/no-multi-comp
    class ContextProvider extends React.Component {
      static childContextTypes = {
        testBroadcast: PropTypes.object,
      };

      getChildContext() {
        return { testBroadcast: broadcast };
      }

      render() {
        return this.props.children;
      }
    }

    const WrappedComponent = ({ testBroadcast }) => <div>{testBroadcast}</div>;

    const GeneratedComponent = hoc(WrappedComponent);

    const tree = renderer
      .create(
        <ContextProvider>
          <GeneratedComponent />
        </ContextProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  // TODO: Add a test to check if the wrapped component is updated when the broadcast updates.
  // This can be done when Enzyme adds support for React 16.
});
