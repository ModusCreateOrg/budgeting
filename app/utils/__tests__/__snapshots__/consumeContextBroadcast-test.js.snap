// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`consumeContextBroadcast should return a HOC that renders correctly if the right broadcast exists in context 1`] = `
<div>
  test
</div>
`;

exports[`consumeContextBroadcast should return a HOC that sets a Broadcast state in the context 1`] = `
<div>
  initialState
</div>
`;
